### PR TITLE
feat: call "cleanup" automatically

### LINKED ISSUE #234: Automatic unmount of rendered components at end of each test
## Describe the Feature
I'm testing hooks that modify the global state of a library, so they need to be unmounted before the next test begins. Currently, I'm using this workaround:
```ts
import { render as renderImpl, RenderAPI } from 'react-native-testing-library'

const rendered: RenderAPI[] = []
const render: typeof renderImpl = (...args) => {
  const elem = renderImpl(...args)
  rendered.push(elem)
  return elem
}

beforeEach(() => {
  rendered.forEach(elem => elem.unmount())
  rendered.length = 0
})
```

Ideally, `react-native-testing-library` would export a function that lets developers opt-in to this behavior, like the following:

```ts
import { setAutoUnmount } from 'react-native-testing-library'

setAutoUnmount(true)
``` 

Then I wouldn't need to maintain this code myself. :)

## Possible Implementations
See the workaround.

## Related Issues
None.

### PR DESCRIPTION
**Pull request by aleclarson**
### Summary

https://github.com/callstack/react-native-testing-library/issues/234#issuecomment-555725623
Fixes #234

Taken from here:

https://github.com/testing-library/react-testing-library/blob/14670debd45236d2c5d0a61a83dadc72af1bef7c/src/index.js

Notice how they also flush the microtask queue.

### Test plan

None needed?

### CODE REVIEW

**Comment by aleclarson** [comment]
In RTL, the `cleanup` function will flush microtasks once [this PR](https://github.com/testing-library/react-testing-library/pull/519) is merged. We should consider doing the same.

**Comment by thymikee** [comment]
Cool, I'll try to get back to this tomorrow :)

**Comment by mdjastrzebski** [comment]
Changes target to `next` branch (which will be ver. 2.0 `master`).

File: src/__tests__/auto-cleanup.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// This just verifies that by importing RNTL in an
+// environment which supports afterEach (like jest)
+// we'll get automatic cleanup between tests.
+test('first', () => {
+  const fn = jest.fn();
+  render(<Test onUnmount={fn} />);
+  expect(fn).not.toHaveBeenCalled();
+});
+
+test('second', () => {
```
Comment (thymikee) [inline src/__tests__/auto-cleanup.js:32]:
```suggestion
test('first - component is mounted, but not umounted before test ends', () => {
  const fn = jest.fn();
  render(<Test onUnmount={fn} />);
  expect(fn).not.toHaveBeenCalled();
});

test('second - component is automatically umounted after first test ends', () => {
```

File: src/__tests__/waitFor.test.js [hunk 2]
```diff
@@ -55,6 +55,8 @@ test('waits for element until timeout is met', async () => {
   await expect(
     waitFor(() => getByText('Fresh'), { timeout: 100 })
   ).rejects.toThrow();
+
+  await waitFor(() => getByText('Fresh'));
```
Comment (thymikee) [inline src/__tests__/waitFor.test.js:59]: why adding this?

File: src/index.js [hunk 3]
```diff
@@ -1,18 +1,17 @@
-// @flow
-import act from './act';
-import cleanup from './cleanup';
-import fireEvent from './fireEvent';
 import flushMicrotasksQueue from './flushMicrotasksQueue';
```
Comment (thymikee) [inline src/index.js:1]: missing `@flow` pragma

File: website/docs/API.md [hunk 4]
```diff
@@ -123,6 +123,8 @@ const cleanup: () => void;
 
 Unmounts React trees that were mounted with `render`.
 
+> Please note that this is done automatically if the testing framework you're using supports the `afterEach` global (like mocha, Jest, and Jasmine). If not, you will need to do manual cleanups after each test.
```
Comment (thymikee) [inline website/docs/API.md:126]:
```suggestion
:::info
Please note that this is done automatically if the testing framework you're using supports the `afterEach` global (like mocha, Jest, and Jasmine). If not, you will need to do manual cleanups after each test.
:::
```

File: website/docs/API.md [hunk 5]
```diff
@@ -123,6 +123,8 @@ const cleanup: () => void;
...
+
 For example, if you're using the `jest` testing framework, then you would need to use the `afterEach` hook like so:
 
 ```jsx
```
Comment (thymikee) [inline website/docs/API.md:130]: Looks like the example should import from `/pure` now?

File: src/__tests__/waitFor.test.js (hunk 2 above)
Comment (mdjastrzebski) [reply src/__tests__/waitFor.test.js:59]: In order to prevent mutation warning after test finishes.

Comment (mdjastrzebski) [reply src/__tests__/waitFor.test.js:59]:
```   console.error
      Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
          in BananaContainer
```

File: src/__tests__/auto-cleanup.js (hunk 1 above)
Comment (mdjastrzebski) [reply src/__tests__/auto-cleanup.js:32]: Updated title

File: src/__tests__/waitFor.test.js (hunk 2 above)
Comment (mdjastrzebski) [reply src/__tests__/waitFor.test.js:59]: Added proper comment

File: src/index.js (hunk 3 above)
Comment (mdjastrzebski) [reply src/index.js:1]: Done

**Comment by thymikee** [comment]
@mdjastrzebski feel free to merge once you think it's ready :)

**Pull request merged**